experiments/compositions/composition/pkg/applier/applier.go

```diff
@@ -278,8 +286,98 @@ func (a *Applier) Apply(oldAppliers []*Applier, prune bool) error {
 	return err
 }
 
-func (a *Applier) Wait() (bool, error) {
-	// TODO(barni@): Do we have standard status fields in KCC, ARC, ...
-	// If so we can wait here. Else it is not feasible to implement a reliable wait
-	return true, nil
+func (a *Applier) getReadinessRule(u *unstructured.Unstructured) string {
+	gvk := u.GetObjectKind().GroupVersionKind()
+	for i := range a.readiness {
+		if a.readiness[i].Group == gvk.Group &&
+			a.readiness[i].Version == gvk.Version &&
+			a.readiness[i].Kind == gvk.Kind {
+			return a.readiness[i].Ready
+		}
+	}
+	return ""
+}
+
+// isObjectReady - is the object ready
+func (a *Applier) isObjectReady(u *unstructured.Unstructured) (bool, string, error) {
+	ready := false
+	var err error
+	message := ""
+	key := types.NamespacedName{
+		Name:      u.GetName(),
+		Namespace: u.GetNamespace(),
+	}
+	err = a.client.Client.Get(a.ctx, key, u)
+	if err != nil {
+		return ready, message, err
+	}
+
+	rule := a.getReadinessRule(u)
+	if rule != "" {
+		// Create a CEL engine
+		celEngine, err := cel.NewEngine(u)
+		if err != nil {
+			return ready, message, fmt.Errorf("error creating CEL engine: %w", err)
+		}
+		result, err := celEngine.Eval(rule)
+		if err != nil {
+			return ready, message, fmt.Errorf("error Evaluating expression: %s, %w", rule, err)
+		}
+		resultBool, ok := result.Value().(bool)
+		if !ok {
+			message := fmt.Sprintf("Readiness Rule [%s] not evaluating to bool. result: %v", rule, result.Value())
+			return ready, message, fmt.Errorf("%s, %w", message, err)
+		}
+		message = fmt.Sprintf("Readiness Rule evaluated to %v", resultBool)
+		ready = resultBool
+	} else {
+		// Fall back to kstatus
+		result, err := status.Compute(u)
+		if err == nil {
+			ready = result.Status == status.CurrentStatus
+			message = result.Message
+		}
+	}
+	return ready, message, err
+}
+
+func (a *Applier) AreResourcesReady() (bool, error) {
+	// Check for readiness or progress CEL rules
+	allReady := true
+	for i, resultObj := range a.results.Objects {
+		// Match objects from this applier only.
+		fromCurrentApplier := false
+		for _, applierObj := range a.objects {
+			if applierObj.GroupVersionKind() == resultObj.GVK &&
+				applierObj.GetNamespace() == resultObj.NameNamespace.Namespace &&
+				applierObj.GetName() == resultObj.NameNamespace.Name {
+				fromCurrentApplier = true
+			}
+		}
+		if fromCurrentApplier {
+			if resultObj.IsPruned {
+				continue
+			}
+			u := &unstructured.Unstructured{}
+			u.SetGroupVersionKind(schema.GroupVersionKind{
+				Group:   resultObj.GVK.Group,
+				Version: resultObj.GVK.Version,
+				Kind:    resultObj.GVK.Kind,
+			})
+			u.SetNamespace(resultObj.NameNamespace.Namespace)
+			u.SetName(resultObj.NameNamespace.Name)
+			ready, message, err := a.isObjectReady(u)
```

Comment:
I _thought_ resultObj was the result of the apply, so you can save yourself the Get

Reply:
https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/blob/master/applylib/applyset/results.go#L25 Doesnt have the full object. 
May be we should return it ?

Reply:
With the callback, we dont have to (yet) return the whole obj. Summary health is returned now.